File: cpp/47_permute_unique/permute_unique.cpp
#include <cassert>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
  vector<int> vis;

 public:
  void backtrack(vector<int>& nums, vector<vector<int>>& ans, int idx, vector<int>& perm) {
    if (idx == nums.size()) {
      ans.emplace_back(perm);
      return;
    }
    for (int i = 0; i < (int)nums.size(); ++i) {
      if (vis[i] || (i > 0 && nums[i] == nums[i - 1] && !vis[i - 1])) {
        continue;
      }
      perm.emplace_back(nums[i]);
      vis[i] = 1;
      backtrack(nums, ans, idx + 1, perm);
      vis[i] = 0;
      perm.pop_back();
    }
  }

  vector<vector<int>> permuteUnique(vector<int>& nums) {
    vector<vector<int>> ans;
    vector<int> perm;
    vis.resize(nums.size());
    sort(nums.begin(), nums.end());
    backtrack(nums, ans, 0, perm);
    return ans;
  }
};

int main() {
  vector<tuple<vector<int>, vector<vector<int>>>> tests{
      {{1, 1, 2}, {{1, 1, 2}, {1, 2, 1}, {2, 1, 1}}},
      {{1, 2, 3}, {{1, 2, 3}, {1, 3, 2}, {2, 1, 3}, {2, 3, 1}, {3, 1, 2}, {3, 2, 1}}},
  };

  for (auto& [nums, ans] : tests) {
    assert(Solution().permuteUnique(nums) == ans);
  }
}

File: rust/47_permute_unique/permute_unique.rs
struct Solution;

impl Solution {
    pub fn permute_unique(nums: Vec<i32>) -> Vec<Vec<i32>> {
        fn backtrack(
            nums: &Vec<i32>,
            ans: &mut Vec<Vec<i32>>,
            perm: &mut Vec<i32>,
            vis: &mut Vec<bool>,
            idx: usize,
        ) {
            if idx == nums.len() {
                ans.push(perm.clone());
                return;
            }
            for i in 0..nums.len() {
                if vis[i] || (i > 0 && nums[i] == nums[i - 1] && !vis[i - 1]) {
                    continue;
                }
                perm.push(nums[i]);
                vis[i] = true;
                backtrack(nums, ans, perm, vis, idx + 1);
                vis[i] = false;
                perm.pop();
            }
        }
        let mut nums = nums.clone();
        nums.sort();
        let mut ans = Vec::new();
        let mut perm = Vec::new();
        let mut vis = vec![false; nums.len()];
        backtrack(&nums, &mut ans, &mut perm, &mut vis, 0);
        ans
    }
}

fn main() {
    let tests = vec![
        (
            vec![1, 1, 2],
            vec![vec![1, 1, 2], vec![1, 2, 1], vec![2, 1, 1]],
        ),
        (
            vec![1, 2, 3],
            vec![
                vec![1, 2, 3],
                vec![1, 3, 2],
                vec![2, 1, 3],
                vec![2, 3, 1],
                vec![3, 1, 2],
                vec![3, 2, 1],
            ],
        ),
    ];

    for (nums, ans) in tests {
        assert_eq!(Solution::permute_unique(nums), ans);
    }
}
